substrate-archive-common/src/database.rs

```diff
@@ -0,0 +1,40 @@
+// Copyright 2017-2019 Parity Technologies (UK) Ltd.
+// This file is part of substrate-archive.
+
+// substrate-archive is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// substrate-archive is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with substrate-archive.  If not, see <http://www.gnu.org/licenses/>.
+
+use super::Result;
+use std::{marker::Sized, path::PathBuf};
+
+pub type KeyValuePair = (Box<[u8]>, Box<[u8]>);
+
+pub const NUM_COLUMNS: u32 = 11;
+
+// Archive specific K/V database reader implementation
+pub trait ReadOnlyDB: Send + Sync {
+    /// Read key/value pairs from the database
+    fn get(&self, col: u32, key: &[u8]) -> Option<Vec<u8>>;
+    /// Iterate over all blocks in the database
+    fn iter<'a>(&'a self, col: u32) -> Box<dyn Iterator<Item = KeyValuePair> + 'a>;
```

Comment:
We could get rid of the Box<dyn ...> here as well:

```rust
pub trait ReadOnlyDB: Send + Sync {
    /// Iterator type to be returned
    type Iter: Iterator<Item = KeyValuePair>;
    /// Read key/value pairs from the database
    fn get(&self, col: u32, key: &[u8]) -> Option<Vec<u8>>;
    /// Iterate over all blocks in the database
    fn iter(&self, col: u32) -> Self::Iter;
```

Unfortunately, `kvdb_rocksdb` uses `impl Iterator` as a return type which obfuscates the concrete iterator type it's returning, which breaks defining the associated type on the trait impl. This would be a small thing to fix as a separate PR later on if the signature is fixed upstream.

Reply:
#123